### PR TITLE
Blast jobs polling and storage using React context

### PR DESCRIPTION
**Pull request by azangru**
## Description
This is one of several alternative PRs (see also https://github.com/Ensembl/ensembl-client/pull/718) to enable the following scenario:

1. Submit the blast from
2. Store the result of the submission
  (the response will be of shape `{submissionId: "some_id", jobs: [ { jobId: "ncbi-blah-blah" }, { error: "error message" }, { jobId: "ncbi-blah-blah" } ]})
    - in redux
    - in indexedDB
3. Poll the jobs while their status is `RUNNING`
4. When the status of a job is reported as `FINISHED` or `FAILURE`:
     - update redux state
     - update indexedDB data
     - stop polling

The solution proposed in this PR has extracted the logic for job status polling and for saving and updating blast submissions in indexedDB into a dedicated tools context. To be compared with https://github.com/Ensembl/ensembl-client/pull/718.

### Advantages and disadvantages of the proposed solution

| Advantages  | Disadvantages  |
|---|---|
| No extra dependencies required | The solution required creating a dedicated context component, and therefore is tied to the structure of the React component tree |
| The polling logic can be represented with an async iterator in a rather simple way |  |

### How to read this PR
The specific functions that run when a Blast form is submitted have been extracted into a dedicated React context (see `ToolsContextContainer`). When the user presses the button in the `BlastJobSubmit` component, the `submitBlastForm` function gets called. It submits the form through the `submitBlast` action creator in `blastApiSlice`, waits for the response (of shape `{ submissionId: string; submission: BlastSubmission }`), saves this to indexedDB, and stats polling for the status of individual jobs within the blast submission.

The polling logic (see the `getFinishedJobs` function) is written as an async iterator.

## Related JIRA Issue(s)
https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-1527
https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-1529

## Deployment URL(s)
http://save-blast-submission.review.ensembl.org

### CODE REVIEW

File: src/content/app/tools/blast/state/blast-api/blastApiSlice.ts [hunk 1]
```diff
@@ -19,6 +19,22 @@ import config from 'config';
...
+
+export type BlastSubmissionPayload = {
+  species: Species[];
+  querySequences: string[];
+  parameters: Record<string, string>;
+};
+
+type BlastSubmissionResponse = {
+  submissionId: string;
+  jobs: Array<{
```
Comment (msmanoj) [inline src/content/app/tools/blast/state/blast-api/blastApiSlice.ts:33]: Should this match with the one here?
https://github.com/Ensembl/ensembl-client/pull/719/files#diff-efd6c7c3fa16b35020ea90f17187298bc502e22972a024c3e5347e43be2d32c2R54

File: src/content/app/tools/blast/state/blast-api/blastApiSlice.ts [hunk 2]
```diff
@@ -27,8 +43,44 @@ const blastApiSlice = restApiSlice.injectEndpoints({
...
+        const results = jobs.map((job) => ({
+          jobId: job.jobId as string,
+          status: 'RUNNING',
+          seen: false,
+          data: null
+        }));
+        return {
+          submissionId,
+          submission: {
+            submission: payload,
```
Comment (msmanoj) [inline src/content/app/tools/blast/state/blast-api/blastApiSlice.ts:75]: `submission` within `submission`?

File: src/root/Root.tsx [hunk 3]
```diff
@@ -63,9 +64,11 @@ export const Root = () => {
   return (
     <div className={styles.root}>
       <ErrorBoundary fallbackComponent={GeneralErrorScreen}>
-        <RootMeta />
-        <App />
-        {showPrivacyBanner && <PrivacyBanner closeBanner={closeBanner} />}
+        <ToolsContextContainer>
```
Comment (msmanoj) [inline src/root/Root.tsx:67]: Wouldn't it be sufficient to just wrap the `<App />` by `ToolsContextContainer`?

File: src/content/app/tools/shared/context/ToolsContextContainer.tsx [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+      await updateSavedBlastJob(updatePayload);
+    }
+  };
+
+  const restoreBlastSubmissions = async () => {
+    const submissions = await getAllBlastSubmissions();
+    dispatch(restoreSubmissions(submissions));
+
+    // resume polling for non-finished jobs
+    const jobIds = Object.entries(submissions).flatMap(
```
Comment (msmanoj) [inline src/content/app/tools/shared/context/ToolsContextContainer.tsx:110]: `jobIds` -> `jobs` ?

**Comment by jyothishnt** [comment]
testing the UI, on submit, it requests a wrong url which returns a  500 internal server error
http://save-blast-submission.review.ensembl.org/api/tools/blast/jobs/status/undefined

**Comment by azangru** [comment]
Closing in favour of either https://github.com/Ensembl/ensembl-client/pull/718 or https://github.com/Ensembl/ensembl-client/pull/722

**Pull request closed without merging**